text/0000-rand-crate-redesign.md

```diff
@@ -0,0 +1,1243 @@
+`Default.sample(&mut rng)`. (Note that `struct Default;` is a unit type (like `()`); Rust
+allows objects to be created without any extra syntax: `let x = Default;`.)
+
+#### Simple distributions
+
+Several zero-size structs implementing [`Distribution`] specify simple distributions:
+
+*   [`Uniform`] specifies uniform distribution over the entire range available, and
+    is implemented for all integer types and `bool`
+*   [`Uniform01`] specifies uniform distribution over the half-open range `[0, 1)`,
+    and is implemented for `f32` and `f64`
+*   [`Closed01`] is like [`Uniform01`] but for `[0, 1]` (thus including 1.0)
+*   [`Open01`] is like [`Uniform01`] but for `(0, 1)` (thus excluding 0.0)
+*   [`Default`] uses [`Uniform`] or [`Uniform01`] depending on type (and can be
+    extended for other types)
+*  [`AsciiWordChar`] samples uniformly from the ASCII characters 0-9, A-Z and a-z
+
+[`Default`] has roughly the same capabilities as the the current `rand` crate's
+[`Rand`](https://docs.rs/rand/0.3.15/rand/trait.Rand.html); currently it doesn't
+support arrays, tuples, `Option`, etc., but support for thes could conceivably
+be added, and probably also an equivalent to `derive_rand`.
+
+It should be noted that there is no agreement on using the name `Default`. In
+particular, there is a naming conflict with `std::default::Default`, which can
+lead to surprising compiler messages if the user forgets to
+`use rand::Default;`.
+Potentially `Default` could be renamed to `Rand`; my personal feeling is that
+the name `Default` works well.
+
+Similarly, `Uniform` and `Uniform01` are open to
+adjustment. All three (including `Default`) could be replaced with a single
+`Uniform`; but using three names does solve
+two semantic issues: (1) the range of sampled values differs by type, especially
+between integer and floating-point types, and (2) some possible
+type-dependent implementations (such as for `Option`) cannot practically have
+a uniform distribution.
+
+[`AsciiWordChar`] is currently an oddity, used in many tests but with no hard
+requirements on form or function. This could be renamed and augmented in line
+with [Regex character classes](https://en.wikipedia.org/wiki/Regular_expression#Character_classes).
+
+#### Range
+
+There is one further uniform distribution:
+
+*   [`Range`] specifies uniform distribution over a range `[a, b)` and supports
+    integer and floating-point types
+
+This [`Range`] is minimally changed from the current `rand`, and supports
+extension to user-defined types by exposing its internal fields.
+
+An alternative
+implementation, [`range2`], has been written in an attempt to improve extension
+to other types and avoid the need for an unused `zone` field with float types,
+but has some drawbacks, perhaps most notably that `Range` is parameterised so
+that `Range::new(low, high)` must be replaced with `range(low, high)` or
+`Range::<T>::new(low, high)`.
```

Comment:
This is a bit ugly... But don't the convenience functions take care of everything? Especially if they are prominent in the documentation, my vote for the `range2` module.

Reply:
Especially since your work, I prefer the `range2` module myself, but you're right about this bit.

I [found a way to allow `Range::new(low, high)`](https://github.com/dhardy/rand/commit/1f1db330e626af805d692dbdcd2dbb982968e533). It's a bit hacky, but works.

Reply:
I have to admit, no idea what is happening here with generic types... But are you sure the design of the library is not depending on a language bug here? Or is it working around a language bug? If so :thumbsup:

Reply:
Nothing very complex... the `new` function is generic in exactly the same way as the `range` function I removed. The `impl Range<T>` bit does nothing at all... that's the "bug" (that a type must be specified even where it's not used).